### DISCUSSION TITLE
Artisan command test doesn't seem to commit created DB record

### DISCUSSION THREAD

**Discussion by LemarinelNet** [laravel/framework#45809] [Q&A]
- Laravel Version: tried with v8.83.27 and v9.48.0
- PHP Version: 8.1.14
- Database Driver & Version: Sqlite3 (SQLite Library => 3.37.2) and MySQL

### Description:
When my Artisan command is run, it creates a model record in the database. die & dumping the table `count()` after the create show one record, but asserting table count in the test gives zero.
Tried with mysql and sqlite : in mysql I could workaround the problem with adding a DB::commit() just after the create (the assert then passes), but in sqlite, it had no effect.

### Steps To Reproduce:
Create a simple artisan command, that tries to `Create` a model : 
```php
<?php

namespace App\Console\Commands;

use App\Models\RegisteredApplication;
use Illuminate\Console\Command;

class RegisterApplication extends Command
{

    protected $signature = 'register:application';
    protected $description = 'Register a new application to the signature service.';

    public function handle()
    {
        $registerd = RegisteredApplication::create();

        return true;
    }
}
```
Create the model and the migration with `php artisan make:model -mf RegisteredApplication`

Create a test that runs this command : 
```php
<?php

namespace Tests\Feature;

use App\Models\RegisteredApplication;
use Tests\TestCase;
use Illuminate\Foundation\Testing\RefreshDatabase;

class ArtisanTest extends TestCase
{
    use RefreshDatabase;

    /** @test  */
    public function ItCreatesAnApplicationRecord()
    {
        $artisan = $this->artisan("register:application");
        $this->assertDatabaseCount("registered_applications",1);
    }
}
```

Run the test : the `assertDatabaseCount` fails.

Github repo to illustrate the issue : `https://github.com/Adesin-fr/LaravelArtisanTest`

PS : as I suspect, it seems to be artisan "mockery" : replacing the line `$artisan = $this->artisan("register:application");` with `Artisan::call("register:application");` makes the test pass, but I can't interact with inputs...

**Comment by eduance** [comment]
Hiya,

Laravel doesn't throw errors by default, but handles em nicely for you. Turn off exception handling in your test: ```$this->withoutExceptionHandling()``` so you can see while your test seems to fail.

You're creating an empty model and haven't added any attributes to your models fillable array. Read more here:

https://laravel.com/docs/9.x/eloquent#inserting-and-updating-models

**Reply by LemarinelNet** [reply]
Hmmm, I'm sorry, but I don't understand something :
Yes, I'm creating an empty model, but NO :
This model has an ID field, and timestamps.

So, I'm not creating an empty model, but a record with just an ID and two timestamps and no extra fields.

And this instruction (mymodel::create() ) is perfectly correct !

Adding $this->withoutExceptionHandling(); doesn't throw any extra errors.

And how would you explain that running Artisan::call instead of this->artisan works ?

PS : Even adding a column to the model and giving it data doesn't solve the issue !
I'm reopening an issue since you've been a bit too comfident when closing it !!!

**Comment by denizgolbas** [comment]
> PS : as I suspect, it seems to be artisan "mockery" : replacing the line `$artisan = $this->artisan("register:application");` with `Artisan::call("register:application");` makes the test pass, but I can't interact with inputs...

You can implement `no-interection` options.

**Comment by mateusjatenee** [comment]
Just stumbled upon this.

The problem is that `artisan` returns an instance of `PendingCommand`, and if you look at that class, it only executes the command when it is destructed (`__destruct()`).  

Since you're referencing it in a variable, it isn't being destructed _prior_ to your DB assertion. Just don't assign it to a variable and you should be good to go.

**Reply by gammamatrix** [reply]
Thanks!

I was assigning `$result = $this->artisan(...`

I removed `$result = ` and it worked for me, and `assertDatabaseHas()` worked as expected.
